packages/backend/src/accounting/psql/ledger-transfer/index.ts

```diff
@@ -38,3 +72,156 @@ export async function getAccountTransfers(
     { credits: [], debits: [] } as GetTransfersResult
   )
 }
+
+export async function createTransfers(
+  deps: ServiceDependencies,
+  transfers: CreateTransferArgs[],
+  trx?: TransactionOrKnex
+): Promise<CreateTransfersResult> {
+  try {
+    const validationResults = await Promise.all(
+      transfers.map((transfer) =>
+        validateTransfer(deps, {
+          creditAccount: transfer.creditAccount,
+          debitAccount: transfer.debitAccount,
+          amount: transfer.amount,
+          timeoutMs: transfer.timeoutMs
+        })
+      )
+    )
+
+    const errors: CreateTransferError[] = []
+
+    for (const [i, maybeError] of validationResults.entries()) {
+      if (isTransferError(maybeError)) {
+        errors.push({
+          index: i,
+          error: maybeError
+        })
+      }
+    }
+
+    if (errors.length > 0) {
+      return {
+        results: [],
+        errors
+      }
+    }
+
+    const createdTransfers = await LedgerTransfer.query(
+      trx || deps.knex
+    ).insertAndFetch(transfers.map(prepareTransfer))
+
+    return {
+      results: createdTransfers,
+      errors: []
+    }
+  } catch (error) {
+    if (error instanceof UniqueViolationError) {
+      return {
+        results: [],
+        errors: [{ index: 0, error: TransferError.TransferExists }]
+      }
+    }
+
+    const errorMessage = 'Could not create transfer(s)'
+    deps.logger.error({ errorMessage: error && error['message'] }, errorMessage)
+    throw new Error(errorMessage)
+  }
+}
+
+async function validateTransfer(
+  deps: ServiceDependencies,
+  args: ValidateTransferArgs
+): Promise<TransferError | undefined> {
+  const { amount, timeoutMs, creditAccount, debitAccount } = args
+
+  if (amount <= 0n) {
+    return TransferError.InvalidAmount
+  }
+
+  if (timeoutMs && timeoutMs < 0n) {
+    return TransferError.InvalidTimeout
+  }
+
+  if (creditAccount.id === debitAccount.id) {
+    return TransferError.SameAccounts
+  }
+
+  if (creditAccount.ledger !== debitAccount.ledger) {
+    return TransferError.DifferentAssets
+  }
+
+  return validateBalances(deps, args)
+}
+
+async function validateBalances(
+  deps: ServiceDependencies,
+  args: ValidateTransferArgs
+): Promise<TransferError | undefined> {
+  const { amount, creditAccount, debitAccount } = args
+
+  const [creditAccountBalances, debitAccountBalance] = await Promise.all([
+    getAccountBalances(deps, creditAccount),
+    getAccountBalances(deps, debitAccount)
+  ])
+
+  if (
+    !hasEnoughDebitBalance({
+      account: creditAccount,
+      balances: creditAccountBalances,
+      transferAmount: amount
+    })
+  ) {
+    return TransferError.InsufficientDebitBalance
+  }
+
+  if (
+    !hasEnoughLiquidity({
+      account: debitAccount,
+      balances: debitAccountBalance,
+      transferAmount: amount
+    })
+  ) {
+    return TransferError.InsufficientLiquidity
+  }
```

Comment:
Should this just be `TransferError.InsufficientBalance`?

Reply:
Since `InsufficientLiquidity` is more referring to the asset liquidity account?

Reply:
I'm not attached to the name, but I think it's still worth having separate errors for `hasEnoughDebitBalance` vs `hasEnoughLiquidity`